File: src/android/SpeedTest.kt
package com.ncinga.speedtest

import android.util.Log
import org.apache.cordova.*
import org.json.JSONArray
import org.json.JSONObject
import com.ookla.speedtest.sdk.SpeedtestSDK

class SpeedTest : CordovaPlugin() {
    private val TAG = "SpeedTest"
    private var customTestHandler: CustomTestHandler? = null
    private var speedtestSDK: SpeedtestSDK? = null

    override fun execute(
        action: String,
        args: JSONArray,
        callbackContext: CallbackContext
    ): Boolean {
        return when (action) {
            "startTesting" -> {
                val jsonObject = parseAndValidateJson(args, callbackContext) ?: return false
                handleStartTesting(jsonObject, callbackContext)
            }

            "stopTesting" -> {
                customTestHandler?.stopTesting()
                return true;
            }

            else -> false
        }
    }

    private fun parseAndValidateJson(
        args: JSONArray,
        callbackContext: CallbackContext
    ): JSONObject? {
        val jsonString = args.optString(0)
        if (jsonString.isNullOrEmpty()) {
            Log.e(TAG, "First argument is not a valid JSON string.")
            callbackContext.error("First argument is not a valid JSON string.")
            return null
        }

        return try {
            JSONObject(jsonString)
        } catch (e: Exception) {
            Log.e(TAG, "Error parsing JSON: ${e.message}")
            callbackContext.error("Error parsing JSON: ${e.message}")
            null
        }
    }

    private fun handleStartTesting(
        jsonObject: JSONObject,
        callbackContext: CallbackContext
    ): Boolean {
        val apiKey = jsonObject.optString("apiKey")
        val config = jsonObject.optString("config")
        val endpoint = jsonObject.optString("endpoint")
        val count = jsonObject.optInt("count", 1)
        val timeInterval = jsonObject.optInt("timeInterval", 10)

        if (apiKey.isEmpty()) {
            Log.e(TAG, "API Key is empty")
            callbackContext.error("API Key is required")
            return false
        }
        if (config.isEmpty()) {
            Log.e(TAG, "Config is required")
            callbackContext.error("Config is required")
            return false
        }
        if (endpoint.isEmpty()) {
            Log.e(TAG, "Endpoint is required")
            callbackContext.error("Endpoint is required")
            return false
        }

        cordova.activity.runOnUiThread {
            try {
                speedtestSDK = SpeedtestSDK.initSDK(cordova.activity.application, apiKey)
                customTestHandler = CustomTestHandler(
                    speedtestSDK!!,
                    config,
                    timeInterval,
                    endpoint,
                    callbackContext
                )
                Log.i(TAG, "Config: $config, Endpoint: $endpoint, Count: $count")
                customTestHandler?.runTestWithSingleServer(count)
            } catch (e: Exception) {
                Log.e(TAG, "Error initializing SpeedtestSDK: ${e.message}")
                callbackContext.error("Error initializing SpeedtestSDK: ${e.message}")
            }
        }
        return true
    }


}

File: src/ios/SpeedTest.swift
import SpeedtestSDK
import os

@objc(SpeedTest)
class SpeedTest: CDVPlugin {
    
    var sdk: STSpeedtestSDK?
    var customTestHandler: CustomTestHandler?
    
    func initializeSDK(apiKey: String) {
        os_log("Initializing SDK...")
        sdk = STSpeedtestSDK.shared
        do {
            try sdk?.initSDK(apiKey)
        } catch {
            os_log("Failed to initialize SDK: %@", "\(error)")
        }
    }
    
    func sendError(message: String, callbackId: String) {
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: message)
        self.commandDelegate.send(pluginResult, callbackId: callbackId)
    }

    @objc(startTesting:)
    func startTesting(command: CDVInvokedUrlCommand) {
        
        guard let jsonString = command.arguments.first as? String,
              let jsonData = jsonString.data(using: .utf8) else {
            sendError(message: "First argument is not a valid JSON string.", callbackId: command.callbackId)
            return
        }
        
        do {
            guard let jsonObject = try JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any] else {
                sendError(message: "Invalid JSON format.", callbackId: command.callbackId)
                return
            }
            
            guard let apiKey = jsonObject["apiKey"] as? String, !apiKey.isEmpty else {
                sendError(message: "API Key is required", callbackId: command.callbackId)
                return
            }
            
            guard let config = jsonObject["config"] as? String, !config.isEmpty else {
                sendError(message: "Config is required", callbackId: command.callbackId)
                return
            }
            
            guard let endpoint = jsonObject["endpoint"] as? String, !endpoint.isEmpty else {
                sendError(message: "Endpoint is required", callbackId: command.callbackId)
                return
            }
            
            let count = jsonObject["count"] as? Int32 ?? 1

            let timeInterval = jsonObject["timeInterval"] as? Int32 ?? 10

            if sdk == nil {
                initializeSDK(apiKey: apiKey)
            }

            guard let sdk = sdk else {
                sendError(message: "SDK initialization failed.", callbackId: command.callbackId)
                return
            }

            customTestHandler = CustomTestHandler(speedTestsdk: sdk, configName: config, count: count, endpoint: endpoint, timeInterval: timeInterval) { result in
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: result)
                print(result)
                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            }

            customTestHandler?.runTestWithSingleServer()


        } catch {
            sendError(message: "Error parsing JSON: \(error.localizedDescription)", callbackId: command.callbackId)
        }
    }
}
